README
ratbox-services
---------------

ratbox-services is a services daemon for use with ircd-ratbox.  It provides
channel/username registration services, nickname services linked in through
usernames, as well as a jupe service, an oper service, an operbot, a 
list service and an oper ban service.


Database backend
----------------
ratbox-services supports a number of database backends for storage, however
it also maintains a large amount of state in RAM instead of querying the
database.  See doc/database_mod.txt for more information.


*********************************************************
* MOST TABLES MUST NOT BE MODIFIED WHILST IT IS RUNNING *
*********************************************************


Installation
------------
Installation instructions are contained in INSTALL.


Upgrading
---------
Important notes when upgrading are contained in UPGRADING.


Release notes
-------------
Notes for this release are contained within RELEASE_NOTES.


Website
-------
Additional documentation is available via the website at:
http://services.ircd-ratbox.org/


Support
-------
Support is available via IRC on #ratbox@EFNet or via email to
ircd-ratbox@lists.ratbox.org.  Please note, if you encounter problems with
ratbox-services, you should *first* look at your logfiles.


Bugs
----
If you encounter a bug with ratbox-services, please report it via email.  If
ratbox-services cores, please keep a copy of the corefile and the binary
safe, and load it into gdb:
gdb /path/to/ratbox-services /path/to/ratbox-services.core

Once at a gdb prompt, type "bt" and email the output.

# $Id$ 


ChangeLog
androsyn    2012/09/04 21:53:34 UTC	(20120904_1-27407)
  Log:
  fix some format type warnings flagged by clang

  Modified:
	ircd-ratbox/trunk/include/messages.h (File Modified) 


androsyn    2012/09/04 21:43:56 UTC	(20120904_0-27405)
  Log:
  suppress Wempty-body warning with clang, this empty body is quite intended

  Modified:
	ircd-ratbox/trunk/src/match.c (File Modified) 


androsyn    2012/07/06 23:13:20 UTC	(20120706_0-27403)
  Log:
  Test commit...
  

  Modified:
	rserv/trunk/INSTALL (File Modified) 


jilles      2012/07/03 23:47:50 UTC	(20120703_0-27401)
  Log:
  tools: Remove ircd-shortcut.pl from the rserv tree.
  
  The version from the ircd tree should be used.
  

  Modified:
	rserv/trunk/INSTALL (File Modified) 
	rserv/trunk/Makefile.in (File Modified) 
	rserv/trunk/tools/ircd-shortcut.pl (File Deleted) 


dubkat      2012/03/16 21:08:18 UTC	(20120316_0-27379)
  Log:
  House Cleaning - update our copyrights

  Modified:
	rserv/trunk/LICENSE (File Modified) 
	rserv/trunk/doc/dbhook.txt (File Modified) 
	rserv/trunk/doc/example.conf (File Modified) 
	rserv/trunk/doc/rsdb.txt (File Modified) 
	rserv/trunk/include/balloc.h (File Modified) 
	rserv/trunk/langs/bg.lang (File Modified) 
	rserv/trunk/langs/example.lang (File Modified) 
	rserv/trunk/langs/genexample.pl (File Modified) 
	rserv/trunk/langs/ru.lang (File Modified) 
	rserv/trunk/src/c_error.c (File Modified) 
	rserv/trunk/src/c_message.c (File Modified) 
	rserv/trunk/src/c_mode.c (File Modified) 
	rserv/trunk/src/cache.c (File Modified) 
	rserv/trunk/src/channel.c (File Modified) 
	rserv/trunk/src/client.c (File Modified) 
	rserv/trunk/src/conf.c (File Modified) 
	rserv/trunk/src/crypt.c (File Modified) 
	rserv/trunk/src/dbhook.c (File Modified) 
	rserv/trunk/src/email.c (File Modified) 
	rserv/trunk/src/hook.c (File Modified) 
	rserv/trunk/src/io.c (File Modified) 
	rserv/trunk/src/langs.c (File Modified) 
	rserv/trunk/src/langs_format.c (File Modified) 
	rserv/trunk/src/log.c (File Modified) 
	rserv/trunk/src/match.c (File Modified) 
	rserv/trunk/src/messages.c (File Modified) 
	rserv/trunk/src/modebuild.c (File Modified) 
	rserv/trunk/src/notes.c (File Modified) 
	rserv/trunk/src/rsdb_mysql.c (File Modified) 
	rserv/trunk/src/rsdb_pgsql.c (File Modified) 
	rserv/trunk/src/rsdb_sqlite3.c (File Modified) 
	rserv/trunk/src/rserv.c (File Modified) 
	rserv/trunk/src/s_alis.c (File Modified) 
	rserv/trunk/src/s_banserv.c (File Modified) 
	rserv/trunk/src/s_chanfix.c (File Modified) 
	rserv/trunk/src/s_chanserv.c (File Modified) 
	rserv/trunk/src/s_global.c (File Modified) 
	rserv/trunk/src/s_jupeserv.c (File Modified) 
	rserv/trunk/src/s_memoserv.c (File Modified) 
	rserv/trunk/src/s_nickserv.c (File Modified) 
	rserv/trunk/src/s_operbot.c (File Modified) 
	rserv/trunk/src/s_operserv.c (File Modified) 
	rserv/trunk/src/s_userserv.c (File Modified) 
	rserv/trunk/src/s_watchserv.c (File Modified) 
	rserv/trunk/src/scommand.c (File Modified) 
	rserv/trunk/src/service.c (File Modified) 
	rserv/trunk/src/tools.c (File Modified) 
	rserv/trunk/src/u_stats.c (File Modified) 
	rserv/trunk/src/ucommand.c (File Modified) 
	rserv/trunk/tools/dbupgrade.pl (File Modified) 
	rserv/trunk/tools/definetolength.pl (File Modified) 
	rserv/trunk/tools/generate-schema.pl (File Modified) 
	rserv/trunk/tools/ircd-shortcut.pl (File Modified) 
	rserv/trunk/tools/sqlitecheck.pl (File Modified) 


moggie      2011/12/20 23:46:02 UTC	(20111220_0-27337)
  Log:
  Avoid using lcase() twice during one function call.
  

  Modified:
	rserv/trunk/src/s_chanfix.c (File Modified) 


moggie      2011/11/14 14:48:32 UTC	(20111114_0-27327)
  Log:
  - Replace SQL's LOWER() for rserv's an IRC style lcase(), LOWER() may cause problems.
  - Fix some bitmask SQL syntax.
  - Use gmtime() instead of localtime().
  - Fix the PgSQL schema.
  - In del_chanfix(), check cfptr in case part_service() causes the channel to be destroyed.
  - Improve reliability of history collation.
  - Fix some SQL queries to conform with stricter syntax.
  - Respond to a server TIME command.
  

  Modified:
	rserv/trunk/src/notes.c (File Modified) 
	rserv/trunk/src/s_banserv.c (File Modified) 
	rserv/trunk/src/s_chanfix.c (File Modified) 
	rserv/trunk/src/s_operbot.c (File Modified) 
	rserv/trunk/src/s_operserv.c (File Modified) 
	rserv/trunk/src/scommand.c (File Modified) 
	rserv/trunk/src/tools.c (File Modified) 
	rserv/trunk/tools/base/schema-pgsql.txt (File Modified) 


androsyn    2011/05/12 17:54:29 UTC	(20110512_3-27201)
  Log:
  free server_p and friends if the server disconnects, allowing auto reconnect to work

  Modified:
	rserv/trunk/src/io.c (File Modified) 


androsyn    2011/05/12 15:33:59 UTC	(20110512_2-27199)
  Log:
  Add -fstack-protector macros and enable them in autoconf
  

  Modified:
	rserv/trunk/aclocal.m4 (File Modified) 
	rserv/trunk/configure (File Modified) 
	rserv/trunk/configure.in (File Modified) 
	rserv/trunk/include/setup.h.in (File Modified) 


androsyn    2011/05/12 15:27:33 UTC	(20110512_1-27197)
  Log:
  remove dependency checking for now..it seems a little hosed

  Modified:
	rserv/trunk/Makefile.in (File Modified) 
	rserv/trunk/src/Makefile.in (File Modified) 


androsyn    2011/05/12 15:21:33 UTC	(20110512_0-27195)
  Log:
  Move the autoconf stuff up to the parent directory like normal.
  
  

  Modified:
	rserv/trunk/autoconf/ (File Deleted) 
	rserv/trunk/config.guess (File Added) 
	rserv/trunk/config.sub (File Added) 
	rserv/trunk/configure (File Modified) 
	rserv/trunk/configure.in (File Added) 
	rserv/trunk/include/setup.h.in (File Modified) 
	rserv/trunk/install-sh (File Added) 


moggie      2011/04/05 12:10:15 UTC	(20110405_0-27187)
  Log:
  - Ensure the USERLIST command displays SVC_CF_NODATAFOR.
  - Add SQL queries to delete unused userhost and channel IDs from the
    database at 28 day intervals.
  

  Modified:
	rserv/trunk/include/s_chanfix.h (File Modified) 
	rserv/trunk/src/s_chanfix.c (File Modified) 



src/conf.c
/* src/conf.c
 *   Contains code for parsing the config file
 *
 * Copyright (C) 2003-2008 Lee Hardy <leeh@leeh.co.uk>
 * Copyright (C) 2003-2012 ircd-ratbox development team
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * 1.Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 * 2.Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * 3.The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id$
 */
#include "stdinc.h"
#include "rserv.h"
#include "langs.h"
#include "conf.h"
#include "tools.h"
#include "client.h"
#include "service.h"
#include "io.h"
#include "log.h"

struct _config_file config_file;
rb_dlink_list conf_server_list;
rb_dlink_list conf_oper_list;

rb_dlink_list client_oper_list;

time_t first_time;

extern int yyparse();           /* defined in y.tab.c */
extern char linebuf[];
extern char conffilebuf[BUFSIZE + 1];
int scount = 0;                 /* used by yyparse(), etc */

FILE *conf_fbfile_in;
extern char yytext[];

static void
set_default_conf(void)
{
	config_file.dcc_low_port = 1025;
	config_file.dcc_high_port = 65000;

	config_file.ping_time = 300;
	config_file.reconnect_time = 300;

	config_file.ratbox = 1;
	config_file.allow_stats_o = 1;

	config_file.default_language = lang_get_langcode(LANG_DEFAULT);

	config_file.client_flood_max = 20;
	config_file.client_flood_max_ignore = 30;
	config_file.client_flood_ignore_time = 300;
	config_file.client_flood_time = 60;

	config_file.min_servers = 0;
	config_file.min_users = 0;
	config_file.max_notes = 10;

	config_file.split_oper_time = 600;

	config_file.disable_email = 1;
	config_file.email_number = 15;
	config_file.email_duration = 60;

	config_file.disable_uregister = 0;
	config_file.uregister_time = 60;
	config_file.uregister_amount = 5;
	config_file.uhregister_time = 86400;	/* 1 day */
	config_file.uhregister_amount = 2;
	config_file.uregister_email = 0;
	config_file.uregister_verify = 0;
	config_file.uexpire_time = 2419200;	/* 4 weeks */
	config_file.uexpire_suspended_time = 2419200;	/* 4 weeks */
	config_file.uexpire_unverified_time = 86400;	/* 1 day */
	config_file.uexpire_bonus_regtime = 4838400;	/* 8 weeks */
	config_file.uexpire_bonus = 86400;		/* 1 day */
	config_file.uexpire_bonus_per_time = 1209600;	/* 2 weeks */
	config_file.uexpire_bonus_max = 2419200;	/* 4 weeks */
	config_file.allow_set_password = 1;
	config_file.allow_resetpass = 0;
	config_file.allow_resetemail = 0;
	config_file.uresetpass_duration = 86400;	/* 1 day */
	config_file.uresetemail_duration = 86400;    /*1 day */
	config_file.ureset_regtime_duration = 1209600; /* 2 weeks */
	config_file.allow_set_email = 1;
	config_file.umax_logins = 5;
	config_file.ushow_suspend_reasons = 0;

	config_file.disable_cregister = 0;
	config_file.cregister_time = 60;
	config_file.cregister_amount = 5;
	config_file.chregister_time = 86400;	/* 1 day */
	config_file.chregister_amount = 4;
	config_file.cexpire_time = 2419200; 	/* 4 weeks */
	config_file.cexpire_suspended_time = 2419200; 	/* 4 weeks */
	config_file.cmax_bans = 50;
	config_file.cexpireban_frequency = 900;		/* 15 mins */
	config_file.cenforcetopic_frequency = 3600;	/* 1 hour */
	config_file.cdelowner_duration = 86400;	/* 1 day */
	config_file.cemail_delowner = 0;
	config_file.cautojoin_empty = 0;
	config_file.cshow_suspend_reasons = 0;
	config_file.ctime_lock = 0;

	config_file.nmax_nicks = 2;
	config_file.nallow_set_warn = 1;

	config_file.os_allow_die = 1;

	config_file.bs_unban_time = 1209600;	/* 2 weeks */
	config_file.bs_temp_workaround = 0;
	config_file.bs_autosync_frequency = DEFAULT_AUTOSYNC_FREQUENCY;
	config_file.bs_regexp_time = 86400;	/* 1 day */
	config_file.bs_merge_into_operserv = 0;
	config_file.bs_max_kline_matches = 200;
	config_file.bs_max_xline_matches = 200;
	config_file.bs_max_resv_matches = 200;
	config_file.bs_max_regexp_matches = 200;

	rb_free(config_file.nwarn_string);
	config_file.nwarn_string = rb_strdup("This nickname is registered, you may "
				"be disconnected if a user regains this nickname.");

	config_file.oper_score = 3;
	config_file.admin_score = 3;
	config_file.jupe_score = 15;
	config_file.unjupe_score = 15;
	config_file.pending_time = 1800;
	config_file.js_merge_into_operserv = 0;

	config_file.ws_merge_into_operserv = 0;

	config_file.ms_max_memos = 50;
	config_file.ms_memo_regtime_duration = 604800;	/* 1 week */

	config_file.max_matches = 60;

	/* chanfix */
	config_file.cf_enable_autofix = 0;
	config_file.cf_enable_chanfix = 0;
	config_file.cf_num_top_scores = 10;
	config_file.cf_min_clients = 4;
	config_file.cf_client_needs_ident = 1;
	config_file.cf_client_needs_dns = 0;
}

static void
validate_conf(void)
{
	if(EmptyString(config_file.name))
		die(0, "No servername specified");

	if(EmptyString(config_file.sid))
		die(0, "No SID specified");

	if(EmptyString(config_file.gecos))
		config_file.gecos = rb_strdup("ratbox services");

	if(config_file.dcc_low_port <= 1024)
		config_file.dcc_low_port = 1025;

	if(config_file.dcc_high_port < config_file.dcc_low_port)
		config_file.dcc_high_port = 65000;

	if(config_file.ping_time <= 0)
		config_file.ping_time = 300;

	if(config_file.reconnect_time <= 0)
		config_file.reconnect_time = 300;

	if(config_file.pending_time <= 0)
		config_file.pending_time = 1800;

	if(config_file.max_matches >= 250)
		config_file.max_matches = 250;
	else if(config_file.max_matches <= 0)
		config_file.max_matches = 250;

	if(config_file.umax_logins < 0)
		config_file.umax_logins = 0;

	if((config_file.max_notes < 0)
			|| (config_file.max_notes > 25))
		config_file.max_notes = 10;

	/* email verification requires we're given an email address */
	if(config_file.uregister_verify)
		config_file.uregister_email = 1;

	/* chanfix validators */
	if((config_file.cf_num_top_scores < 1)
			|| (config_file.cf_num_top_scores > 10))
		config_file.cf_num_top_scores = 10;

	if(config_file.cf_min_clients < 2)
		config_file.cf_min_clients = 2;

}

static void
clear_old_conf(void)
{
	struct conf_oper *oper_p;
	rb_dlink_node *ptr;
	rb_dlink_node *next_ptr;
	int i;

	for(i = 0; config_file.email_program[i]; i++)
	{
		rb_free(config_file.email_program[i]);
		config_file.email_program[i] = NULL;
	}

	RB_DLINK_FOREACH_SAFE(ptr, next_ptr, conf_oper_list.head)
	{
		oper_p = ptr->data;

		/* cleaned up properly in clear_old_conf_final() */
		SetConfDead(oper_p);
	}

	RB_DLINK_FOREACH_SAFE(ptr, next_ptr, conf_server_list.head)
	{
		free_conf_server(ptr->data);
		rb_dlinkDestroy(ptr, &conf_server_list);
	}	
}

/* clear_old_conf_final()
 *   Clears out old configuration stuff that rserv doesn't need after a
 *   rehash.
 */
static void
clear_old_conf_final(void)
{
	struct conf_oper *oper_p;
	rb_dlink_node *ptr;
	rb_dlink_node *next_ptr;

	RB_DLINK_FOREACH_SAFE(ptr, next_ptr, conf_oper_list.head)
	{
		oper_p = ptr->data;

		if(ConfDead(oper_p) && oper_p->refcount == 0)
		{
			free_conf_oper(oper_p);
			rb_dlinkDestroy(ptr, &conf_oper_list);
		}
	}
}

void
conf_parse(int cold)
{
	struct client *target_p;
	rb_dlink_node *ptr;

        if((conf_fbfile_in = fopen(CONF_PATH, "r")) == NULL)
	{
		if(!cold)
		{
			mlog("Failed to open config file");
			sendto_all("Failed to open config file");
			return;
		}
		else
	                die(0, "Failed to open config file");
	}

	if(!cold)
		clear_old_conf();
	else
		set_default_conf();

        yyparse();
	validate_conf();

	/* if we havent sent our burst, the following will just break */
	if(!testing_conf && sent_burst)
	{
		RB_DLINK_FOREACH(ptr, service_list.head)
		{
			target_p = ptr->data;

			if(ServiceIntroduced(target_p))
			{
				if(ServiceDisabled(target_p))
				{
					deintroduce_service(target_p);
					continue;
				}
				else if(ServiceReintroduce(target_p))
				{
					reintroduce_service(target_p);
					continue;
				}
			}
			else if(!ServiceDisabled(target_p))
				introduce_service(target_p);

			ClearServiceReintroduce(target_p);
		}
	}

        fclose(conf_fbfile_in);

	if(!cold)
		clear_old_conf_final();
}

void
rehash(int sig)
{
	if(sig)
	{
		mlog("services rehashing: got SIGHUP");
		sendto_all("services rehashing: got SIGHUP");
	}

	reopen_logfiles();

	conf_parse(0);
}

void
free_conf_oper(struct conf_oper *conf_p)
{
	rb_free(conf_p->name);
	rb_free(conf_p->pass);
	rb_free(conf_p->username);
	rb_free(conf_p->host);
	rb_free(conf_p->server);
	rb_free(conf_p);
}

void
free_conf_server(struct conf_server *conf_p)
{
	rb_free(conf_p->name);
	rb_free(conf_p->host);
	rb_free(conf_p->pass);
	rb_free(conf_p->vhost);
}

void
deallocate_conf_oper(struct conf_oper *conf_p)
{
	conf_p->refcount--;

	/* marked as dead, now unused, free. */
	if(ConfDead(conf_p) && !conf_p->refcount)
		free_conf_oper(conf_p);
}

struct conf_server *
find_conf_server(const char *name)
{
        struct conf_server *server;
        rb_dlink_node *ptr;

        RB_DLINK_FOREACH(ptr, conf_server_list.head)
        {
                server = ptr->data;

                if(!strcasecmp(name, server->name))
                        return server;
        }

        return NULL;
}

struct conf_oper *
find_conf_oper(const char *username, const char *host, const char *server,
		const char *oper_username)
{
        struct conf_oper *oper_p;
        rb_dlink_node *ptr;

        RB_DLINK_FOREACH(ptr, conf_oper_list.head)
        {
                oper_p = ptr->data;

		if(ConfDead(oper_p))
			continue;

                if(match(oper_p->username, username) &&
                   match(oper_p->host, host) &&
		   (EmptyString(oper_p->server) || match(oper_p->server, server)) &&
		   (EmptyString(oper_username) || !irccmp(oper_p->name, oper_username)))
                        return oper_p;
        }

        return NULL;
}

/* store_client_oper()
 *   Stores an opers current logged in information
 *
 * inputs	- client to store information for
 * outputs	-
 * side effects - adds an entry to client_oper_list
 */
void
store_client_oper(struct client *client_p)
{
	struct client_oper *coper_p;

	coper_p = rb_malloc(sizeof(struct client_oper));
	rb_strlcpy(coper_p->uid, client_p->uid, sizeof(coper_p->uid));
	rb_strlcpy(coper_p->username, client_p->user->username,
		sizeof(coper_p->username));
	rb_strlcpy(coper_p->host, client_p->user->host, sizeof(coper_p->host));
	rb_strlcpy(coper_p->sid, client_p->uplink->uid, sizeof(coper_p->sid));

	coper_p->oper = client_p->user->oper;
	coper_p->oper->refcount++;

	coper_p->timestamp = rb_time();

	rb_dlinkAdd(coper_p, &coper_p->ptr, &client_oper_list);
}

/* free_client_oper()
 *   frees a client_oper and removes from list
 *
 * inputs	- struct client_oper to free
 * outputs	-
 * side effects	- frees the client_oper and removes from client_oper_list
 */
void
free_client_oper(struct client_oper *coper_p)
{
	deallocate_conf_oper(coper_p->oper);
	rb_dlinkDelete(&coper_p->ptr, &client_oper_list);
	rb_free(coper_p);
}

/* clear_client_oper_sid()
 *   clears all client oper entries for a given SID
 *
 * inputs	- server to clear from client_oper_list
 * outputs	-
 * side effects - clears all entries for server from client_oper_list
 */
void
clear_client_oper_sid(struct client *client_p)
{
	struct client_oper *coper_p;
	rb_dlink_node *ptr, *next_ptr;

	RB_DLINK_FOREACH_SAFE(ptr, next_ptr, client_oper_list.head)
	{
		coper_p = ptr->data;

		if(!irccmp(coper_p->sid, client_p->uid))
			free_client_oper(coper_p);
	}
}

/* expire_client_oper()
 *   expires entries from client_oper_list
 *
 * inputs	- NULL
 * outputs	-
 * side effects - expires stale entries from client_oper_list
 */
void
expire_client_oper(void *unused)
{
	struct client_oper *coper_p;
	rb_dlink_node *ptr, *next_ptr;

	RB_DLINK_FOREACH_SAFE(ptr, next_ptr, client_oper_list.head)
	{
		coper_p = ptr->data;

		if((coper_p->timestamp + config_file.split_oper_time) < rb_time())
			free_client_oper(coper_p);
	}
}

static struct flag_table
{
	char mode;
	int flag;
} oper_flags[] = {
	{ 'D', CONF_OPER_DCC		},
	{ 'A', CONF_OPER_ADMIN		},
	{ '\0',0 }
};

const char *
conf_oper_flags(unsigned int flags)
{
	static char buf[20];
	static const char *empty_flags = "-";
	char *p = buf;
	int i;

	for(i = 0; oper_flags[i].mode; i++)
	{
		if(flags & oper_flags[i].flag)
			*p++ = oper_flags[i].mode;
	}

	*p = '\0';

	if(EmptyString(buf))
		return empty_flags;

	return buf;
}

#undef SERVICE_FLAGS_FULL

static struct sflag_table
{
	int serviceid;
	const char *sname;
	uint64_t flag;
	const char *name;
} service_flags[] = {
#ifdef SERVICE_FLAGS_FULL
	{ 1, "operserv", CONF_OPER_OS_MAINTAIN, "maintain"	},
	{ 1, "operserv", CONF_OPER_OS_IGNORE,	"ignore"	},
	{ 1, "operserv", CONF_OPER_OS_CHANNEL,	"channel"	},
	{ 1, "operserv", CONF_OPER_OS_TAKEOVER,	"takeover"	},
	{ 1, "operserv", CONF_OPER_OS_OMODE,	"omode"		},
	{ 2, "userserv", CONF_OPER_US_REGISTER,	"register"	},
	{ 2, "userserv", CONF_OPER_US_SUSPEND,	"suspend"	},
	{ 2, "userserv", CONF_OPER_US_DROP,	"drop"		},
	{ 2, "userserv", CONF_OPER_US_SETPASS,	"setpass"	},
	{ 2, "userserv", CONF_OPER_US_LIST,	"list"		},
	{ 2, "userserv", CONF_OPER_US_INFO,	"info"		},
	{ 3, "chanserv", CONF_OPER_CS_REGISTER,	"register"	},
	{ 3, "chanserv", CONF_OPER_CS_SUSPEND,	"suspend"	},
	{ 3, "chanserv", CONF_OPER_CS_DROP,	"drop"		},
	{ 3, "chanserv", CONF_OPER_CS_LIST,	"list"		},
	{ 3, "chanserv", CONF_OPER_CS_INFO,	"info"		},
	{ 4, "nickserv", CONF_OPER_NS_DROP,	"drop"		},
	{ 5, "operbot",  CONF_OPER_OB_CHANNEL,	"channel"	},
	{ 6, "global",   CONF_OPER_GLOB_NETMSG,	"netmsg"	},
	{ 6, "global",   CONF_OPER_GLOB_WELCOME,"welcome"	},
	{ 7, "jupeserv", CONF_OPER_JS_JUPE,	"jupe"		},
	{ 8, "banserv",  CONF_OPER_BAN_KLINE,	"kline"		},
	{ 8, "banserv",  CONF_OPER_BAN_XLINE,	"xline"		},
	{ 8, "banserv",  CONF_OPER_BAN_RESV,	"resv"		},
	{ 8, "banserv",  CONF_OPER_BAN_REGEXP,	"regexp"	},
	{ 8, "banserv",  CONF_OPER_BAN_PERM,	"perm"		},
	{ 8, "banserv",  CONF_OPER_BAN_REMOVE,	"remove"	},
	{ 8, "banserv",  CONF_OPER_BAN_SYNC,	"sync"		},
	{ 8, "banserv",	 CONF_OPER_BAN_NOMAX,	"nomax"		},
	{ 9, "chanfix",  CONF_OPER_CF_INFO,	"info"		},
	{ 9, "chanfix",  CONF_OPER_CF_NOTES,	"notes"		},
	{ 9, "chanfix",  CONF_OPER_CF_BLOCK,	"block"	},
	{ 9, "chanfix",  CONF_OPER_CF_CHANFIX,	"chanfix"	},
#else
	{ 1, "OS", CONF_OPER_OS_MAINTAIN,	"M" },
	{ 1, "OS", CONF_OPER_OS_IGNORE,		"I" },
	{ 1, "OS", CONF_OPER_OS_CHANNEL,	"C" },
	{ 1, "OS", CONF_OPER_OS_TAKEOVER,	"T" },
	{ 1, "OS", CONF_OPER_OS_OMODE,		"O" },
	{ 2, "US", CONF_OPER_US_REGISTER,	"R" },
	{ 2, "US", CONF_OPER_US_SUSPEND,	"S" },
	{ 2, "US", CONF_OPER_US_DROP,		"D" },
	{ 2, "US", CONF_OPER_US_SETPASS,	"P" },
	{ 2, "US", CONF_OPER_US_LIST,		"L" },
	{ 2, "US", CONF_OPER_US_INFO,		"I" },
	{ 3, "CS", CONF_OPER_CS_REGISTER,	"R" },
	{ 3, "CS", CONF_OPER_CS_SUSPEND,	"S" },
	{ 3, "CS", CONF_OPER_CS_DROP,		"D" },
	{ 3, "CS", CONF_OPER_CS_LIST,		"L" },
	{ 3, "CS", CONF_OPER_CS_INFO,		"I" },
	{ 4, "NS", CONF_OPER_NS_DROP,		"D" },
	{ 5, "OB", CONF_OPER_OB_CHANNEL,	"C" },
	{ 6, "GL", CONF_OPER_GLOB_NETMSG,	"N" },
	{ 6, "GL", CONF_OPER_GLOB_WELCOME,	"W" },
	{ 7, "JS", CONF_OPER_JS_JUPE,		"J" },
	{ 8, "BS", CONF_OPER_BAN_KLINE,		"K" },
	{ 8, "BS", CONF_OPER_BAN_XLINE,		"X" },
	{ 8, "BS", CONF_OPER_BAN_REGEXP,	"P" },
	{ 8, "BS", CONF_OPER_BAN_RESV,		"R" },
	{ 8, "BS", CONF_OPER_BAN_PERM,		"P" },
	{ 8, "BS", CONF_OPER_BAN_REMOVE,	"V" },
	{ 8, "BS", CONF_OPER_BAN_SYNC,		"S" },
	{ 8, "BS", CONF_OPER_BAN_NOMAX,		"M" },
	{ 9, "CF", CONF_OPER_CF_INFO,		"I" },
	{ 9, "CF", CONF_OPER_CF_NOTES,		"N" },
	{ 9, "CF", CONF_OPER_CF_BLOCK,		"B" },
	{ 9, "CF", CONF_OPER_CF_CHANFIX,	"C" },
#endif
	{ 0, NULL, 0, NULL }
};

const char *
conf_service_flags(uint64_t flags)
{
	static char buf[BUFSIZE];
	static const char *empty_flags = "-";
	int i;
	int serviceid = 0;

	buf[0] = '\0';

	for(i = 0; service_flags[i].flag; i++)
	{
		if(flags & service_flags[i].flag)
		{
			if(serviceid != service_flags[i].serviceid)
			{
				if(serviceid)
					rb_strlcat(buf, " ", sizeof(buf));

				rb_strlcat(buf, service_flags[i].sname, sizeof(buf));
				rb_strlcat(buf, ":", sizeof(buf));

				serviceid = service_flags[i].serviceid;
			}
#ifdef SERVICE_FLAGS_FULL
			else
				rb_strlcat(buf, ",", sizeof(buf));
#endif
			rb_strlcat(buf, service_flags[i].name, sizeof(buf));
		}
	}

	if(EmptyString(buf))
		return empty_flags;

	return buf;
}

/*
 * yyerror
 *
 * inputs	- message from parser
 * output	- none
 * side effects	- message to opers and log file entry is made
 */
void
yyerror(const char *msg)
{
	char newlinebuf[BUFSIZE];

	strip_tabs(newlinebuf, (const unsigned char *) linebuf, strlen(linebuf));

	sendto_all("\"%s\", line %d: %s at '%s'",
                   conffilebuf, lineno + 1, msg, newlinebuf);

	mlog("conf error: \"%s\", line %d: %s at '%s'", conffilebuf, lineno + 1, msg, newlinebuf);
}

int
conf_fbgets(char *lbuf, int max_size)
{
	char *buff;

	if((buff = fgets(lbuf, max_size, conf_fbfile_in)) == NULL)
		return (0);

	return (strlen(lbuf));
}
